### PR TITLE
Better driving controller

### PR DESCRIPTION
**Pull request by DJSteins**
Adds ramp support and modifies the input from keys to steering to work how it did with the old driving controller.

### CODE REVIEW

File: Assets/Script/Game/Car/CarController.cs [hunk 1]
```diff
@@ -343,13 +371,26 @@ private void AccelerateCar()
         m_box.AddForce(acceleration * Time.deltaTime, ForceMode.VelocityChange);
     }
 
+    private void RotateToGround()
+    {
+        // reference: https://answers.unity.com/questions/168097/orient-vehicle-to-ground-normal.html?_ga=2.154419966.778972075.1583965656-1799704818.1555470483
+        RaycastHit hit;
+        if (Physics.Raycast(m_box.transform.position, -m_curNormal, out hit))
+        {
+            m_curNormal = Vector3.Lerp(m_curNormal, hit.normal, 4 * Time.deltaTime);
```
Comment (m4rtint) [inline Assets/Script/Game/Car/CarController.cs]: The last time i saw you testing, was when you rotated the floor.  (So the Rotation of the floor is not Quart.identity)
Have you tried testing with an actual ramp object where the mesh is zeroed out? Does that give you the correct normal?

File: Assets/Script/Game/Car/CarController.cs [hunk 2]
```diff
@@ -273,17 +277,41 @@ private Vector3 GetCenterOfRotation(float radius, float omega, float turnDirecti
         return wheelPosition + radius * (Quaternion.AngleAxis(turnDirection * 90.0f, Vector3.up) * forward);
     }
 
+    private float GetCameraTurnModifier()
+    {
+        if (!m_isHolding)
+        {
+            return 1;
+        }
+        float dt = Time.time - m_horizontalHeldDown;
```
Comment (m4rtint) [inline Assets/Script/Game/Car/CarController.cs]: what is the horizontalHeldDown for? why is it related to time?

File: Assets/Script/Game/Car/CarController.cs [hunk 3]
```diff
@@ -55,6 +55,10 @@ public class CarController : MonoBehaviour
     /// Use this timer to override the movement of the car
     /// </summary>
     private float m_movementOverrideTimer;
+    Vector3 m_curNormal = Vector3.up;
+    float m_horizontalHeldDown = 0.0f;
+    bool m_isHolding = false;
+    float m_lastSignal = 1;
```
Comment (Gursaanj) [inline Assets/Script/Game/Car/CarController.cs]: Missing Protection Levels

File: Assets/Script/Game/Car/CarController.cs (hunk 1 above)
Comment (DJSteins) [reply Assets/Script/Game/Car/CarController.cs]: Actual ramp object? If you look at the video I put in the coding channel the ramp there is a plane with a box collider on it and it works fine on that ramp.

File: Assets/Script/Game/Car/CarController.cs (hunk 2 above)
Comment (DJSteins) [reply Assets/Script/Game/Car/CarController.cs]: It's for checking how long the right or left directional input is held down. This is to handle the case where the driver is holding down left or right while the direction of the car changes from facing towards the camera to facing away from the camera or vise-versa.

File: Assets/Script/Game/Car/CarController.cs (hunk 1 above)
Comment (m4rtint) [reply Assets/Script/Game/Car/CarController.cs]: Mmm those ramps are rotate to a certain degrees.
If you were to use ramps which are rotation 0,0,0, would you get the same outcome?

**Comment by m4rtint** [comment]
PLease merge from develop

File: Assets/Script/Game/Car/CarController.cs (hunk 1 above)
Comment (m4rtint) [reply Assets/Script/Game/Car/CarController.cs]: Tested this out, it works np

**Comment by m4rtint** [comment]
Bug: I'm not sure if it's related to the ramps. 
When you hold down forward and press left/right/ over the ramps?? 
and then u switch directions, it will go the wrong direction.

Hold forward + hold right
go over ramp
Keep holding forward + hold Left.

Car still goes right. 

Hard to pin point  exact repo steps, but easy to do it randomly

**Comment by mikeslgoh** [comment]
[Refactor car controller to test with a ramp](https://trello.com/c/KZNUQQls/411-refactor-car-controller-to-test-with-a-ramp)

**Pull request merged**